feat(modal): set incremental z-index value for modals and asides to allow for stacking

DaleWebb opened the pull request:
feat(modal): set incremental z-index value for modals and asides to allow for stacking

The number of active modals and asides are tracked in order to set a `z-index` value that will allow those elements to be on top of each other.
Currently, if two asides are shown at the same time, clicking the backdrop will hide the aside that is below the one above.  Also, the backdrop of the second aside is below the first aside.

mgcrea commented on src/modal/modal.js:
```diff
@@ -28,6 +28,9 @@ angular.module('mgcrea.ngStrap.modal', ['mgcrea.ngStrap.core', 'mgcrea.ngStrap.h
       var bodyElement = angular.element($window.document.body);
 
+      var modalCount = 0;
+      var baseZindexValue = 1037;
```
Any reason you picked this specific number? Looks like it should default to 1050 according to https://github.com/twbs/bootstrap/blob/master/less/variables.less#L277

DaleWebb replied:
I used the default value for the backdrop set here previously, which was
1038. Then took advantage of the fact that the element is one of the latter
child elements of the body to state that it is on top.

Should I set the modal z-index to start at 1050?
On 5 Nov 2015 23:24, "Olivier Louvignes" <notifications@github.com> wrote:

> In src/modal/modal.js
> <https://github.com/mgcrea/angular-strap/pull/1946#discussion_r44084925>:
>
> > @@ -28,6 +28,9 @@ angular.module('mgcrea.ngStrap.modal', ['mgcrea.ngStrap.core', 'mgcrea.ngStrap.h
> >        var requestAnimationFrame = $window.requestAnimationFrame || $window.setTimeout;
> >        var bodyElement = angular.element($window.document.body);
> >
> > +      var modalCount = 0;
> > +      var baseZindexValue = 1037;
>
> Any reason you picked this specific number? Looks like it should default
> to 1050 according to
> https://github.com/twbs/bootstrap/blob/master/less/variables.less#L277
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mgcrea/angular-strap/pull/1946/files#r44084925>.
>

mgcrea commented:
We should closely follow bootstrap variables to avoid breaking people apps. Also we should manage separate z-indexes for backdrop and modal, first one being 1040/1050 (same as bootstrap), second one 1060/1070, etc. So that you still can easily manage items in-between layers.

DaleWebb commented:
Okay, so first modal/aside has the backdrop set at 1040 and dialog set at
1050, then the second backdrop being 1060 and dialog being 1070 as so on?
On 6 Nov 2015 09:34, "Olivier Louvignes" <notifications@github.com> wrote:

> We should closely follow bootstrap variables to avoid breaking people
> apps. Also we should manage separate z-indexes for backdrop and modal,
> first one being 1040/1050 (same as bootstrap), second one 1060/1070, etc.
> So that you still can easily manage items in-between layers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mgcrea/angular-strap/pull/1946#issuecomment-154360355>
> .
>

mgcrea commented:
Yep.

DaleWebb commented:
I've added unit tests about the z-indexes for each `.aside` and `.modal` and adjusted the starting value of the `z-index` to follow the bootstrap convention.  In a previous commit, this pull-request affected the other components that inherited `$modal` e.g. `$alert`, so now the `z-index` modifications will only happen if the component has a backdrop.

mgcrea commented on src/modal/modal.js:
```diff
@@ -184,6 +197,9 @@ angular.module('mgcrea.ngStrap.modal', ['mgcrea.ngStrap.core', 'mgcrea.ngStrap.h
           if(!$modal.$isShown) return;
 
+          // decrement number of modals
+          backdropCount--;
```
should be wrapped in a `if(options.backdrop)`.

DaleWebb commented:
Updated

mgcrea commented:
Thanks! Merged :+1:
